sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -154,6 +154,17 @@ object ScalaReflection extends ScalaReflection {
     case _ => "array"
   }
 
+  private[sql] def defaultValueFor(enc: AgnosticEncoder[_]): Any = enc match {
+      case PrimitiveBooleanEncoder => null.asInstanceOf[Boolean]
+      case PrimitiveByteEncoder => null.asInstanceOf[Byte]
+      case PrimitiveShortEncoder => null.asInstanceOf[Short]
+      case PrimitiveIntEncoder => null.asInstanceOf[Int]
+      case PrimitiveLongEncoder => null.asInstanceOf[Long]
+      case PrimitiveFloatEncoder => null.asInstanceOf[Float]
+      case PrimitiveDoubleEncoder => null.asInstanceOf[Double]
+      case _ => null
```

Comment:
It's a bit confusing to call it as `defaultValue`?  Usually for Boolean, default value is `false`, etc.

This is more like `nullValueFor`?

Reply:
`null.asInstanceOf[Boolean]` will return false. The problem with ReduceAggregator was the type T was considered of type Encoder, so even if T is a Boolean during runtime, we receive `b = null`, which causes the NPE. So I have to enforce the cast here during compilation based on the encoder to correct the default value.

Reply:
Ha nice to know. Thanks!